go/service/chat_local.go

```diff
@@ -89,58 +175,133 @@ func (h *chatLocalHandler) CompleteAndCanonicalizeTlfName(ctx context.Context, t
-// keybase.chatLocal.GetOrCreateTextConversationLocal protocol. It returns the
-// most recent conversation's ConversationID for given TLF ID, or creates a new
-// conversation and returns its ID if none exists yet.
-//
-// TODO: after we implement multiple conversations per TLF and topic names,
-// change this to look up by topic name
-//
-// TODO: cache ConversationIDs and conversations in service
-func (h *chatLocalHandler) ResolveConversationLocal(ctx context.Context, arg keybase1.ConversationInfoLocal) (ids []chat1.ConversationID, err error) {
-	res, err := h.boxer.tlf.CryptKeys(ctx, arg.TlfName)
-	if err != nil {
-		return ids, err
+// keybase.chatLocal.GetOrCreateTextConversationLocal protocol.
+func (h *chatLocalHandler) ResolveConversationLocal(ctx context.Context, arg keybase1.ConversationInfoLocal) (conversations []keybase1.ConversationInfoLocal, err error) {
+	if arg.Id != 0 {
+		info, _, err := h.getConversationInfo(ctx, arg.Id)
+		if err != nil {
+			return nil, err
+		}
+		return []keybase1.ConversationInfoLocal{info}, nil
 	}
-	tlfIDb, err := hex.DecodeString(string(res.TlfID))
-	if err != nil {
-		return ids, err
+	return h.searchForConversations(ctx, arg)
+}
+
+// searchForConversations searches for conversations using tlfName, topicName,
+// and topicType fields in criteria, and returns all matching conversations.
+// Conversation IDs are populated in returned conversations.
+func (h *chatLocalHandler) searchForConversations(ctx context.Context, criteria keybase1.ConversationInfoLocal) (conversations []keybase1.ConversationInfoLocal, err error) {
+	appendMaybe := func(info keybase1.ConversationInfoLocal) {
+		if len(criteria.TopicName) > 0 && criteria.TopicName != info.TopicName {
+			return
+		}
+		if criteria.TopicType != chat1.TopicType_NONE && criteria.TopicType != info.TopicType {
+			return
+		}
+		conversations = append(conversations, info)
 	}
-	tlfID := chat1.TLFID(tlfIDb)
 
-	ipagination := &chat1.Pagination{Num: 20}
-getinbox:
-	for i := 0; i < 10000; /* in case we have a server bug */ i++ {
-		iview, err := h.GetInboxLocal(ctx, ipagination)
+	if len(criteria.TlfName) > 0 { // fast case
+		// TODO: do some caching in boxer so we don't end up calling this RPC
+		// unnecessarily too often
+		resp, err := h.boxer.tlf.CryptKeys(ctx, criteria.TlfName)
 		if err != nil {
-			return ids, err
+			return nil, err
 		}
-		for _, conv := range iview.Conversations {
-			if conv.Metadata.IdTriple.Tlfid.Eq(tlfID) {
-				// TODO: check topic name and topic ID here when we support multiple
-				// topics per TLF
-				ids = append(ids, conv.Metadata.ConversationID)
+		tlfIDb, err := hex.DecodeString(string(resp.TlfID))
+		if err != nil {
+			return nil, err
+		}
+		conversationsRemote, err := h.remoteClient().GetInboxByTLFIDRemote(ctx, tlfIDb)
+		if err != nil {
+			return nil, err
+		}
+		for _, cr := range conversationsRemote {
+			info, _, err := h.getConversationInfo(ctx, cr.Metadata.ConversationID)
```

Comment:
We already have the `Conversation` here, can we just pass that into `getConversationInfo`? Otherwise that guy is just fetching it from the server again. Is the plan just to cache that, or can we just be smarter about the parameters that `getConversationInfo` takes?